streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -475,28 +466,43 @@ void closeNonAssignedSuspendedTasks(final Map<TaskId, Set<TopicPartition>> newAs
                     suspendedTask.closeSuspended(true, false, null);
                 } catch (final Exception e) {
                     log.error("Failed to remove suspended {} {} due to the following error:", taskTypeName, suspendedTask.id(), e);
-                } finally {
-                    standByTaskIterator.remove();
                 }
+                standByTaskIterator.remove();
             }
         }
     }
 
     void close(final boolean clean) {
-        close(allTasks(), clean);
-        clear();
-    }
-
-    private void close(final Collection<Task> tasks, final boolean clean) {
-        for (final Task task : tasks) {
+        for (final Task task : allTasks()) {
             try {
                 task.close(clean, false);
-            } catch (final Throwable t) {
+            } catch (final ProducerFencedException ignoreAndSwallow) {
+                closeZombieTask(task);
+            } catch (final Throwable logAndSwallow) {
                 log.error("Failed while closing {} {} due to the following error:",
                           task.getClass().getSimpleName(),
                           task.id(),
-                          t);
+                    logAndSwallow);
+                closeUncleanIfRequired(task, clean);
             }
         }
+
+        clear();
+    }
+
+    private void closeUncleanIfRequired(final Task task,
```

Comment:
Could we inline it inside `close`, or is there any specific reasons for dragging it out as a separate function?

Reply:
It was inlined originally, and requested by @bbejeck to put into it's own method to make code more readable.